### ISSUE TITLE
Add support for running script-sandbox in external machine

### ISSUE THREAD

**Issue by putsi** [swissfondue/gtta#13]
Currently, GTTA runs the sandbox-vm locally with OpenVZ. This makes it impossible to run scripts in environments, where the GTTA-machine does not have direct access to internet/targets.

Would it be possible/easy to add support for using OpenVZ server that is located on an external machine (see the image below)?

![kuva](https://cloud.githubusercontent.com/assets/5388424/26400828/95e5350e-408b-11e7-89b0-10cb4dafa54a.PNG)

**Comment by swissfondue** [comment]
Hi Jarmo

 

Wouldnt it be easier to use GTTA locally on a laptop (slave), run all the tools and then create a master/slave setup that would allow a slave to sync the data then later with a master? This is something we thought of anyway at one point

 

Best

Oliver

 

Von: Jarmo Puttonen [mailto:notifications@github.com] 
Gesendet: Mittwoch, 24. Mai 2017 13:18
An: swissfondue/gtta <gtta@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Betreff: [swissfondue/gtta] Add support for running script-sandbox in external machine (#13)

 

Currently, GTTA runs the sandbox-vm locally with OpenVZ. This makes it impossible to run scripts in environments, where the GTTA-VM does not have direct access to internet/targets.

Would it be possible/easy to add support for using OpenVZ server that is located on an external machine (see the image below)?

 <https://cloud.githubusercontent.com/assets/5388424/26400828/95e5350e-408b-11e7-89b0-10cb4dafa54a.PNG>

**Comment by anton-belousov** [comment]
GTTA uses VMManager to control the OpenVZ container, but it utilises the only implementation based on vzctl shell command, which works only with the local container. I'm not sure if OpenVZ has  a remote control capability, if it has, then this function could be implemented easily by modifying the VMManager class.

Otherwise we will need to install some additional software on the OpenVZ side. Probably some simple API endpoint or something like that.

**Comment by putsi** [comment]
Alright I'll look into both ideas (master/slave and remote contorl OpenVZ) and try figure out some solution.
Will do a pull request if I manage to create some clean and simple solution.

**Comment by putsi** [comment]
I managed to do a pretty clean implementation which modifies only VMManager.

If the user specifies in VMManager.php const variables that a remote OpenVZ-server (e.g. [Virtuozzo](https://openvz.org/Virtuozzo)) should be used, the VMManager will send vzctl-commands through SSH-connection to the external server. The VMManager also used some filesystem-commands, however that was solved by mounting the related directory of the external OpenVZ-server through sshfs.

I'll send a pull request for this soon. As usual, feel free to reject the pull request if you think that there is no need for such feature.
